### ISSUE TITLE
Compiling in release with gcc 8.2 causes FP math issues

### ISSUE THREAD

**Issue by olafklingt** [supercollider/supercollider#4116]
Environment
-----------

* Your SuperCollider version: 3.9.3 (installed from the comunity repositories of arch)
* Your operating system and version: Arch Linux 

Steps to reproduce (for bugs)
-----------------------------

```supercollider
(0/0).isNaN
// Please paste SuperCollider code here.
// It really helps if you try to simplify your example as much as possible.
```

Expected Behavior
-----------------
returns true

Current Behavior
----------------
returns false

isNaN is implemented by testing < and > of 0  ... and these tests are already not behaving as expected.

**Comment by brianlheim** [comment]
> and these tests are already not behaving as expected.

What are the results of `0/0`, `0/0 > 0`, and `0/0 < 0` on your machine? Can you please give your CPU/architecture?

**Comment by olafklingt** [comment]
(0/0)  -> -nan
(0/0) > 0 -> true
(0/0) < 0 -> false
(0/0) == 1 -> true
(0/0) == -1 -> true
(0/0) === 1 -> false
(0/0) === (0/0) -> true

my cpu is a recent intel one.  i run this linux as a 64bit system.

when i look at the supercollider about field i see the following line:
"Built from branch 'not_a_git_checkout' [na]"
so when this bug is rare maybe the compiling for arch linux has not been done correctly?



less /proc/cpuinfo output is:

processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 142
model name      : Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
stepping        : 9
microcode       : 0x8e
cpu MHz         : 3271.939
cache size      : 4096 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf
bogomips        : 5810.00
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual

**Comment by brianlheim** [comment]
> "Built from branch 'not_a_git_checkout' [na]"
so when this bug is rare maybe the compiling for arch linux has not been done correctly?

You could ask @dvzrv, he is the maintainer for that Arch package.

**Comment by brianlheim** [comment]
It sounds like gcc is compiling Supercollider with `-ffast-math`.

**Comment by brianlheim** [comment]
It is indeed: https://github.com/supercollider/supercollider/blob/develop/CMakeLists.txt#L110

That's unfortunate. Can someone verify on Arch that applying this patch and recompiling solves the issue:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 58a8c80da..782ed5cc9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -107,7 +107,7 @@ if (${CMAKE_COMPILER_IS_GNUCXX})
                message(FATAL_ERROR "SuperCollider requires at least gcc-4.8 when compiled with gcc.")
        endif()

-       add_definitions("-ffast-math -fsigned-zeros -fno-associative-math")
+       add_definitions("-fno-math-errno -fno-signaling-nans -fsigned-zeros -fno-associative-math")

        if(APPLE)
                exec_program(${CMAKE_CXX_COMPILER} ARGS --version OUTPUT_VARIABLE _gcc_version_info)
```

**Comment by brianlheim** [comment]
Also @olafklingt if you don't mind humoring me further, what does the following produce on your system?

```supercollider
a = (0/0);
a.high32Bits.asBinaryString(32) + a.low32Bits.asBinaryString(32)
```

Here it is `-> 11111111111110000000000000000000 00000000000000000000000000000000`

**Comment by olafklingt** [comment]
here it is:
-> 11111111111110000000000000000000 00000000000000000000000000000000

I need some days until i find time to compile and contact more people. but i will come back to it.

**Comment by brianlheim** [comment]
OK! this could be behavior from a newer version of gcc or maybe it does have to do with something going wrong during compilation... i doubt it is any specific CPU issue but you never know. i used to have an Arch VM but recently lost it, so i can't check myself right now.

**Comment by dvzrv** [comment]
Hm, can have a look into this later or tomorrow!

**Comment by brianlheim** [comment]
Thanks @dvzrv !

**Comment by dvzrv** [comment]
@brianlheim  I can confirm, that your change to the gcc flags fixes the issue!
I have rolled out a new revision of the package on Arch!

**Comment by brianlheim** [comment]
@dvzrv thanks! I am not sure yet whether this is a good fix and something we should push to the main branch. Honestly I'm surprised that `-ffast-math` worked as long as it did without running into something like this; we should be respecting floating point math a little bit more than this...

also, just to be clear, you were able to reproduce this in a build before using the patch above?

**Comment by olafklingt** [comment]
The new build solves the problem for me, thanks!

**Comment by brianlheim** [comment]
Great! I am going to relabel this issue, it should stay open until we decide what to do in mainline.

**Comment by olafklingt** [comment]
just found this 
https://stackoverflow.com/questions/22931147/stdisinf-does-not-work-with-ffast-math-how-to-check-for-infinity
which refers to here
http://gcc.gnu.org/onlinedocs/gcc-4.8.2/gcc/Optimize-Options.html#Optimize-Options
which highlights that fast-math is implying -ffinite-math-only which means:
"Allow optimizations for floating-point arithmetic that assume that arguments and results are not NaNs or +-Infs.
This option is not turned on by any -O option since it can result in incorrect output for programs that depend on an exact implementation of IEEE or ISO rules/specifications for math functions. It may, however, yield faster code for programs that do not require the guarantees of these specifications."

**Comment by brianlheim** [comment]
@olafklingt yep, I'm aware of the advertised behavior of -ffast-math, which is why I made the suggestion in the first place. But, suddenly changing a bunch of floating point math switches for everyone who compiles with GCC (i.e. all Linux users) may have other unintended effects. I don't know which side to err on, and right now we haven't even fully determined what combinations of hardware and compiler this issue is present for.

With no better hunches to go on, my current inclination is this:

- change the flags for gcc >= 8.2 only, for the 3.10 release
- explicitly ask that Linux users test this when we put out the release candidate, and adjust accordingly
- after releasing 3.10, in the development branch (i.e., for 3.11), make the full switch as given in my patch above

**Comment by dvzrv** [comment]
@brianlheim yes, the package revision before the applied fix had this issue on my machine as well. I'm not sure since when.
We still have gcc7  (7.3.1) in the repos. I could try a test compilation with it the next days to see if it has the same issues (or are you building against that as well?).

**Comment by dvzrv** [comment]
The problem is also reproducible with supercollider 3.9.3, compiled using gcc 7.3.1.
Guess this has been around for much longer (which is bad).

**Comment by brianlheim** [comment]
@dvzrv interesting! can anyone reproduce this on other repos? (@snappizz , @patrickdupuis ) i would have expected this to be reported sooner if it was always the case with gcc 7 and 8. maybe it has something to do with the gcc package on that distribution? that seems unlikely.

**Comment by patrickdupuis** [comment]
On Fedora 28, gcc 8.1.1:
```
(0/0).isNaN // false
(0/0) // -nan
(0/0) > 0 // true
(0/0) < 0 // false
(0/0) == 1 // true
(0/0) == -1 // true
(0/0) === 1 // false
(0/0) === (0/0) // true

a = (0/0);  // -nan
a.high32Bits.asBinaryString(32) + a.low32Bits.asBinaryString(32)
// 11111111111110000000000000000000 00000000000000000000000000000000
```

**Comment by olafklingt** [comment]
Just a thought. Does the c side of sclang make use of testing for nan for proper operation? Otherwise this could be also considered a problem which can be solved on the sclang side by re-implementing isNaN and > < >= <= ... or the primitives for it.

**Comment by brianlheim** [comment]
> Does the c side of sclang make use of testing for nan for proper operation?

What do you mean? Also sclang is written in C++, not C.

> Otherwise this could be also considered a problem which can be solved on the sclang side by re-implementing isNaN and > < >= <= ... or the primitives for it.

It's not even a primitive in this case, it's a single bytecode that executes the C++ expression `x op y`, where `op` is the appropriate comparison operator. (ref: https://github.com/supercollider/supercollider/blob/develop/lang/LangSource/PyrMathOps.cpp#L446-L451)

**Comment by brianlheim** [comment]
@patrickdupuis Thanks, so let's definitely prioritize #4119 for 3.10. If you have time to run the master test suite for that build please do so.

**Comment by olafklingt** [comment]
> > Does the c side of sclang make use of testing for nan for proper operation?
> 
> What do you mean? Also sclang is written in C++, not C.

I mean C++, sorry.

> 
> > Otherwise this could be also considered a problem which can be solved on the sclang side by re-implementing isNaN and > < >= <= ... or the primitives for it.
> 
> It's not even a primitive in this case, it's a single bytecode that executes the C++ expression `x op y`, where `op` is the appropriate comparison operator. (ref: https://github.com/supercollider/supercollider/blob/develop/lang/LangSource/PyrMathOps.cpp#L446-L451)

I mean that these lines could be overwritten with customized functions. like it is done for the mod operator:
https://github.com/supercollider/supercollider/blob/f543f534d516b6d7a798577ea403a661497a792a/lang/LangSource/PyrMathOps.cpp#L753

But only when inside sclang's c++ implementation a correct interpretation of nan's is not important as well.

but after thinking a bit longer about it i do not think that this can be more efficient than not using fast-math. But I simply don't know enough.

**Comment by patrickdupuis** [comment]
For good mesure, on Ubuntu 18.04, GCC 7.3.0:

```
(0/0) // -nan
(0/0) > 0 // true
(0/0) < 0 // false
(0/0) == 1 // true
(0/0) == -1 // true
(0/0) === 1 // false
(0/0) === (0/0) // true
a = (0/0) // -nan
a.high32Bits.asBinaryString(32) + a.low32Bits.asBinaryString(32)
// 11111111111110000000000000000000 00000000000000000000000000000000
```

**Comment by patrickdupuis** [comment]
Bonus: Rpi 3, Raspbian Stretch, gcc 6.3.0

```
(0/0).isNaN // true
(0/0) // nan
(0/0) > 0 // false
(0/0) < 0 // false
(0/0) == 1 // false
(0/0) == -1 // false
(0/0) === 1 // false
(0/0) === (0/0) // true
a = (0/0) // nan
a.high32Bits.asBinaryString(32) + a.low32Bits.asBinaryString(32)
// 01111111111110000000000000000000 00000000000000000000000000000000
```

The results are the same when building #4119 and current `3.10`. Unfortunately, there isn't an easy way to get gcc 7+ on Debian Stretch.

**Comment by brianlheim** [comment]
Awesome, thanks @patrickdupuis !

**Comment by brianlheim** [comment]
Fix is in 3.10 branch now! Thanks for the testing and report @olafklingt @dvzrv

**Comment by brianlheim** [comment]
OK, was able to produce a minimal example that shows this in action, for pedagogical purposes :)

```cpp
bool op(double a, double b, int c)
{
    return (a / b) == c;
}

int main() {
    return op(0.0, 0.0, 1);
}
```

https://godbolt.org/z/wClLjV

you can see that the generated assembly is definitely not the same; in the non-fast-mathed version the PF flag is being checked (`setnp`, line 7) which is what's set if one of the operands is NaN. essentially those three lines before the `ret` mean:
- `ucomisd`: compare the result of the division with the third argument cast to a double
- `setnp`: set the result to `false` is either argument was NaN, and `true` otherwise
- `cmovne`: set the result to `false` if the arguments were not equal, do nothing otherwise

on the right hand side, you can see the three instructions `ucomisd / setnp / cmovne` have been replaced with `comisd / sete`. if you look at the documentation for [`comisd`](http://www.felixcloutier.com/x86/COMISD.html) and [`sete`](http://www.felixcloutier.com/x86/SETcc.html) you can see that the ZF flag is set by `comisd` if the two arguments are equal, but _also_ if they're unordered (i.e. either is NaN). so, the behavior of this function is:
- compare the result of the division with the third argument cast to a double
- set the result to `true` if the arguments are equal or if either is NaN, `false` otherwise

you can also see on the right hand side that gcc is even able to tell at compile time that the comparison in `f()` is obviously going to return `false`. it can't make that deduction

you can use this tiny program given above to confirm this:

```sh
[brianheim@BrianMBP fastmath]$ cat main.cpp
bool op(double a, double b, int c)
{
    return (a / b) == c;
}

int main() {
    return op(0.0, 0.0, 1);
}
[brianheim@BrianMBP fastmath]$ g++-8 -o main main.cpp
[brianheim@BrianMBP fastmath]$ ./main ; echo $?
0
[brianheim@BrianMBP fastmath]$ g++-8 -o main main.cpp -ffast-math
[brianheim@BrianMBP fastmath]$ ./main ; echo $?
1
```

i couldn't make a similar reproducer for the other bad comparisons shown in this thread `<`, `>`, `<=`, `>=`; not sure why. maybe gcc is optimizing it differently in context, or i'm misunderstanding what's going on in the actual implementation. anyway, hope this helps to demystify the topic!
